google/cloud/datastore/batch.py

```diff
@@ -267,12 +281,19 @@ def rollback(self):
         Marks the batch as aborted (can't be used again).
 
         Overridden by :class:`google.cloud.datastore.transaction.Transaction`.
+
+        :raises: :class:`~exceptions.ValueError` if the batch is not
+                 in progress.
         """
+        if self._status != self._IN_PROGRESS:
+            raise ValueError('Batch must be in progress to rollback()')
+
         self._status = self._ABORTED
 
     def __enter__(self):
-        self._client._push_batch(self)
         self.begin()
+        # NOTE: We make sure begin() succeeds before pushing onto the stack.
+        self._client._push_batch(self)
```

Comment:
The error in #2297 comes from the fact that the new `Batch` instance is pushed on first (`self._client._push_batch(self)`) and then `self.begin()` is called, which is where the API call (via `begin_transaction`) failed. So after that, the `Client`s stack is just stuck with that `Batch`.

The thing that makes it extra "bad" is that if the stack is not empty when it should be, calls like `Client.put()` will use that batch / transaction, instead of creating a new one. And those methods only `commit()` when they create their own `Batch` during the method call.